modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java

```diff
@@ -0,0 +1,325 @@
+            buff.position(offsetOf(recNumber));
+            env.expandToInclude(buff.get(), buff.get());
+            if (!pointBounds) {
+                env.expandToInclude(buff.get(), buff.get());
+            }
+        }
+    }
+
+    /** Uses a heap memory backed {@link DoubleBuffer} */
+    private static class HeapBoundsReader extends BufferBoundsReader {
+        HeapBoundsReader(int recCount, boolean pointBounds) {
+            super(DoubleBuffer.allocate((pointBounds ? 2 : 4) * recCount), pointBounds);
+        }
+    }
+
+    /**
+     * Uses a memory mapped temporary file as the source for the {@link DoubleBuffer} where to store
+     * the record bounds.
+     *
+     * <p>{@link #close()} releases the memory mapped byte buffer and deletes the temporary file.
+     */
+    private static class FileBoundsReader implements BoundsReader {
+
+        private static final String TMP_FILE_NAME_PREFIX = "GeoTools_shp_qix_bounds_";
+        private final BufferBoundsReader reader;
+        private final Path path;
+        private final RandomAccessFile file;
+        private final MappedByteBuffer mappedBuffer;
+
+        FileBoundsReader(final int recCount, final boolean isPoint) throws IOException {
+            final int fileSize = recCount * Double.BYTES * (isPoint ? 2 : 4);
+            path = Files.createTempFile(TMP_FILE_NAME_PREFIX, ".tmp");
+            try {
+                checkAvailableSpace(path, fileSize);
+                file = new RandomAccessFile(path.toFile(), "rw");
+                mappedBuffer = file.getChannel().map(MapMode.READ_WRITE, 0L, fileSize);
+                reader = new BufferBoundsReader(mappedBuffer.asDoubleBuffer(), isPoint);
+            } catch (IOException | RuntimeException e) {
+                Files.delete(path);
+                throw e;
+            }
+        }
+
+        @Override
+        public void close() {
+            reader.close();
+            NIOUtilities.clean(mappedBuffer, true);
+            try {
+                file.close();
+            } catch (IOException e) {
+                LOGGER.log(Level.WARNING, "Error closing temporary file " + path, e);
+            }
+            try {
+                Files.delete(path);
+            } catch (IOException e) {
+                LOGGER.log(Level.WARNING, "Error deleting temporary file " + path, e);
+            }
+        }
+
+        @Override
+        public void insert(int recno, Envelope env) {
+            reader.insert(recno, env);
+        }
+
+        @Override
+        public void read(int recNumber, Envelope env) throws IOException {
+            reader.read(recNumber, env);
+        }
+
+        @Override
+        public void expand(int recNumber, Envelope env) throws IOException {
+            reader.expand(recNumber, env);
+        }
+
+        private void checkAvailableSpace(final Path path, final int fileSize)
+                throws IOException, FileSystemException {
```

Comment:
FileSystemException is a subclass of IOException, so, redundant in the throw.

Reply:
Done.